### PR TITLE
Allow return statements within looping/branching statements in ANTLR grammar

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
This adds the ability for `return` statements within loop (`while/for`) and branch (`if/else`) in the ANTLR grammar.

### CODE REVIEW

File: source/grammar/qasm3.g4 [hunk 1]
```diff
@@ -353,14 +356,17 @@ loopSignature
...
 loopStatement: loopSignature programBlock ;
 
-controlDirectiveStatement
-    : controlDirective SEMICOLON
+endStatement
+    : 'end' SEMICOLON
     ;
 
+returnStatement
+    : 'return' statement;
```
Comment (hhorii) [inline source/grammar/qasm3.g4:364]: I think `statement` should be `statement?` to allow following flow.
```
def sub(int[64] i) {
  if (i < 0) return;
  // do something
}
```

Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4]: Fair pt. Addressed in https://github.com/Qiskit/openqasm/pull/140/commits/9e85ffe8ae570a9769726f6f1a380660beaca65b.

File: source/grammar/qasm3.g4 [hunk 2]
```diff
@@ -220,13 +226,18 @@ quantumInstruction
...
     | quantumMeasurement
+    | quantumReset
     | quantumBarrier
     ;
 
 quantumPhase
     : 'gphase' LPAREN Identifier RPAREN
     ;
 
+quantumReset
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:237]: Seperated bc can't add gate modifiers to `reset`

**Comment by zachschoenfeld33** [comment]
Updated w/ a new test and merged master branch

File: source/language/subroutines.rst [hunk 3]
```diff
@@ -5,10 +5,11 @@ Subroutines are declared using the statement ``def name(parameters) qargs -> out
 Zero or more quantum bits
 and registers are passed to the subroutine by reference or name in ``qargs``.
 Classical types are passed by value in ``parameters``. The parentheses may be omitted if no
-``parameters`` are passed. The subroutines return up to
-one classical type. All arguments are declared together with their type,
-for example ``qubit: ancilla`` would define a quantum bit argument named ``ancilla``. Qubit
-declarations are not allowed within subroutines as they are global. A subroutine
+``parameters`` are passed. The subroutines return up to one classical type, signified by the
```
Comment (taalexander) [inline source/language/subroutines.rst:8]:
```suggestion
``parameters`` are passed. The subroutines return up to one value of a classical type, signified by the
```

Comment (zachschoenfeld33) [reply source/language/subroutines.rst]: Addressed in https://github.com/Qiskit/openqasm/pull/140/commits/f5ea050b6c7c47ec9d236139689023568aebcc99

**Review by taalexander — APPROVED** [review]

**Pull request merged**